src/proof.rs

```diff
@@ -0,0 +1,332 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::nibbleslice::NibbleSlice;
+use crate::node::Node;
+use ccrypto::{blake256, BLAKE_NULL_RLP};
+use primitives::H256;
+
+// Unit of a proof.
+//#[derive(Clone, Eq, PartialEq, Debug, RlpEncodable, RlpDecodable)]
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct CryptoProofUnit<H, K, V> {
+    pub hash: H, // or root
+    pub key: K,
+    pub value: Option<V>, // None in case of absence
+}
+
+// Abstract trait of being cryptographically provable.
+pub trait CryptoProvable<H, K, V> {
+    // Does this verify the unit? Note: Invalid / ill-formed proof should return false, not abort or error
+    fn verify(&self, test: &CryptoProofUnit<H, K, V>) -> bool;
+}
+
+// Abstract trait of being a data structure that can provide CryptoProvable for its element.
+pub trait CryptoStructure<H, K, V> {
+    fn make_proof(&self, key: &K) -> crate::Result<(CryptoProofUnit<H, K, V>, Box<dyn CryptoProvable<H, K, V>>)>;
+}
+
+// This is strongly bound to implementation of TrieDB. I want to specify that but rust has no HKT(?)
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct CryptoProofMerkleTrie {
+    pub proof: Vec<Vec<u8>>, // Starts with the closest node to root.
+}
+
+pub struct CryptoProofSkewedTree {
+    // TODO
+}
+
+impl CryptoProvable<H256, H256, Vec<u8>> for CryptoProofMerkleTrie {
+    // Proof should start with root node.
+    fn verify(&self, test: &CryptoProofUnit<H256, H256, Vec<u8>>) -> bool {
+        type Tunit = CryptoProofUnit<H256, H256, Vec<u8>>;
+        // step1: verify the value
+        fn step1(self_: &CryptoProofMerkleTrie, test: &Tunit) -> bool {
+            match Node::decoded(&self_.proof[self_.proof.len() - 1]) {
+                Some(x) => match x {
+                    Node::Leaf(_, value) => test.value.as_ref().unwrap() == &value,
+                    _ => false,
+                },
+                _ => false,
+            }
+        };
+
+        // step2: verify the root
+        fn step2(self_: &CryptoProofMerkleTrie, test: &Tunit) -> bool {
+            blake256(&self_.proof[0]) == test.hash
+        };
+
+        // step3 (presence): verify the key
+        fn step3_p(self_: &CryptoProofMerkleTrie, test: &Tunit) -> bool {
+            fn verify_branch(path: &NibbleSlice<'_>, hash: &H256, proof: &[Vec<u8>]) -> bool {
+                *hash == blake256(&proof[0])
+                    && match Node::decoded(&proof[0]) {
+                        Some(Node::Leaf(partial, _)) => path == &partial,
+                        Some(Node::Branch(partial, table)) => {
+                            // Note: Does Rust guarantee the short circuit evaluation?
+                            proof.len() >= 2 && // detect ill-formed proof 
```

Comment:
Prefer to use if statement and early return instead.

Reply:
What's the difference?

Reply:
Readability.

Reply:
@sgkim126 Would you provide me a general guideline for this kind of situation? When to `if` and when to `&&`?

Reply:
Oh sorry this page was not refreshed after you comment. I'll use 'if' when it seems to be more readable.